cmd/src/dump.rs

```diff
@@ -29,58 +39,116 @@ pub fn check_and_dump_raft_db<E: RaftEngine>(raftdb_path: &str, engine: &E, env:
-                        _ => unreachable!(),
-                    }
-                    if count_entry % 1024 == 0 {
-                        let size = engine.consume(&mut log_batch, true).unwrap();
-                        count_size += size;
-                    }
-                    Ok(true)
-                }
+    let mut threads = vec![];
+    let (tx, rx) = unbounded();
+    for _ in 0..thread_num {
+        let raft_engine = engine.clone();
+        let raftdb_engine = origin_engine.clone();
+        let count_size = count_size.clone();
+        let rx = rx.clone();
+        let t = std::thread::spawn(move || {
+            run_worker(rx, raftdb_engine, raft_engine, count_size);
+        });
+        threads.push(t);
+    }
+    info!("Start to scan raft log from raftdb and dump into raft engine");
+    let consumed_time = std::time::Instant::now();
+    // Seek all region id from raftdb and send them to workers.
+    let mut it = origin_engine.iterator().unwrap();
+    let mut valid = it.seek(SeekKey::Key(keys::REGION_RAFT_MIN_KEY)).unwrap();
+    while valid {
+        match keys::decode_raft_key(it.key()) {
+            Err(e) => {
+                error!("Error happened when decoding raft key: {}", e);
+                break;
             }
-        })
-        .unwrap();
-    let size = engine.consume(&mut log_batch, true).unwrap();
+            Ok((id, _)) => {
+                tx.send(id).unwrap();
+                count_region += 1;
+                let next_key = keys::raft_log_prefix(id + 1);
+                valid = it.seek(SeekKey::Key(&next_key)).unwrap();
+            }
+        }
+    }
+    drop(tx);
+    info!("Scanned all region id and waiting for dump");
+    for t in threads {
+        t.join().unwrap();
+    }
     engine.sync().unwrap();
-    count_size += size;
     info!(
         "Finished dump, total regions: {}; Total bytes: {}; Consumed time: {:?}",
         count_region,
-        count_size,
-        consume_time.elapsed(),
+        count_size.load(Ordering::Relaxed),
+        consumed_time.elapsed(),
     );
     convert_to_dirty_raftdb(raftdb_path);
     check_and_delete_safely(raftdb_path);
 }
 
+// Worker receives region id and scan the related range.
+fn run_worker(
+    rx: Receiver<u64>,
+    old_engine: RocksEngine,
+    new_engine: RaftLogEngine,
+    count_size: Arc<AtomicUsize>,
+) {
+    let mut batch = new_engine.log_batch(0);
+    let mut local_size = 0;
+    while let Ok(id) = rx.recv() {
+        let mut entries = vec![];
+        old_engine
+            .scan(
+                &keys::raft_log_prefix(id),
+                &keys::raft_log_prefix(id + 1),
+                false,
+                |key, value| {
+                    let res = keys::decode_raft_key(key);
+                    match res {
+                        Err(_) => Ok(true),
+                        Ok((region_id, suffix)) => {
+                            local_size += value.len();
```

Comment:
size should be key+value?

Reply:
The key size is almost negligible, and the local size itself is not accurate statistics